c7n/resources/vpc.py

```diff
@@ -2708,3 +2708,65 @@ class resource_type(query.TypeInfo):
         arn_type = 'traffic-mirror-target'
         universal_taggable = object()
         id_prefix = 'tmt-'
+
+
+@RouteTable.filter_registry.register('cross-az-nat-gateway-route')
+class CrossAZRouteTable(Filter):
+    """Filter route-table which is attached to a subnet that is not in the same availability zone (AZ) of the
+    NAT Gateway subnet AZ. This filter is applicable to those route-tables which have next hop as NatGateways for
+    any route. If a route-table has NAT gateway from us-west-2a and it is associated to subnet from us-west-2b then it
+    will flagged.
+    This filter is useful for cost optimization and performance use-cases, where we don't want network traffic to cross
+    from one availability zone (AZ) to another AZ.
+
+    :Example:
+    .. code-block:: yaml
+            policies:
+              - name: cross-az-nat-gateway-traffic
+                resource: aws.route-table
+                filters:
+                    - type: cross-az-nat-gateway-route
+                actions:
+                  - notify
+    """
+    schema = type_schema('cross-az-nat-gateway-route')
+    permissions = ("ec2:DescribeRouteTables", "ec2:DescribeNatGateways", "ec2:DescribeSubnets")
+
+    def process(self, resources, event=None):
+        # dump of all subnets and nat-gateways to avoid multiple API calls
+        all_subnets = self.manager.get_resource_manager('aws.subnet').resources()
+        all_nat_gws = self.manager.get_resource_manager('nat-gateway').resources()
+        # Create subnet_id AZ mapping
+        subnets_az_map = {subnet["SubnetId"]: subnet["AvailabilityZone"] for subnet in all_subnets}
+        # Create nat_gateway subnet_id AZ mapping vpc wise
+        nat_gws_subnet_map = {nat_gateway['NatGatewayId']: nat_gateway["SubnetId"]
+                              for nat_gateway in all_nat_gws}
+        # List of AZ where NAT gateways are created vpc wise. To make sure NAT Gateway is available in a given AZ for a vpc
+        all_nat_gw_az = {}  # To store NAT Gateway subnet AZ per VPC ID
+        for nat_gateway in all_nat_gws:
+            vpc_id = nat_gateway["VpcId"]
+            subnet_az = subnets_az_map[nat_gateway["SubnetId"]]
+            if vpc_id in all_nat_gw_az:
+                all_nat_gw_az[vpc_id].append(subnet_az)
+            else:
+                all_nat_gw_az[vpc_id] = [subnet_az]
+        results = []  # To return filtered resources
+        for res in resources:
+            vpc_id = res["VpcId"]
+            res['NatGatewayAvailabilityZone'] = {}
```

Comment:
prefix with c7n: when adding custodian attributes/annotations is preferred.

Reply:
Added c7n prefix